pandas/tests/io/test_sql.py

```diff
@@ -577,6 +642,180 @@ def sample(pd_table, conn, keys, data_iter):
         # Nuke table
         self.drop_table("test_frame1")
 
+    def _to_sql_on_conflict_update(self, method, pkey_type):
+        """
+        GIVEN:
+        - Original database table: 3 rows
+        - new dataframe: 4 rows (2 duplicate keys)
+        WHEN:
+        - on conflict update insert
+        THEN:
+        - DB table len = 5
+        - Conflicting primary keys in DB updated
+        """
+        # Nuke
+        self.drop_table("pkey_table")
+        # Re-create original table
+        self._create_pkey_table(pkey_type)
+        # Original table exists and as 3 rows
+        assert self.pandasSQL.has_table("pkey_table")
+        assert self._count_rows("pkey_table") == 3
+        # Insert new dataframe
+        self.pandasSQL.to_sql(
+            self.pkey_table_frame,
+            "pkey_table",
+            if_exists="append",
+            on_conflict="do_update",
+            index=False,
+            method=method,
+        )
+        # Check table len correct
+        assert self._count_rows("pkey_table") == 5
+        # Check conflicting primary keys have been updated
+        # Get new values for conflicting keys
+        duplicate_keys = [1, 2]
+        duplicate_key_query = SQL_STRINGS["read_pkey_table"][self.flavor]
+        duplicate_val = self._get_exec().execute(duplicate_key_query, duplicate_keys)
+        data_from_db = sorted(val[0] for val in duplicate_val)
+        # Expected values from pkey_table_frame
+        expected = sorted(["new_val1", "new_val2"])
+        assert data_from_db == expected
+        # Finally, confirm that duplicate values are not removed from original df object
+        assert len(self.pkey_table_frame.index) == 4
+        # Clean up
+        self.drop_table("pkey_table")
+
+    def _to_sql_on_conflict_nothing(self, method, pkey_type):
+        """
+        GIVEN:
+        - Original table: 3 rows
+        - new dataframe: 4 rows (2 duplicate keys)
+        WHEN:
+        - on conflict do nothing insert
+        THEN:
+        - database table len = 5
+        - conflicting keys in table not updated
+        """
+        # Nuke
+        self.drop_table("pkey_table")
+        # Re-create original table
+        self._create_pkey_table(pkey_type)
+        # Original table exists and has 3 rows
+        assert self.pandasSQL.has_table("pkey_table")
+        assert self._count_rows("pkey_table") == 3
+        # Prepare SQL for reading duplicate keys
```

Comment:
I see the below pattern repeated a few times - possible to extract into a method? (Same for any others)

```
        # Prepare SQL for reading duplicate keys
        duplicate_keys = [1, 2]
        duplicate_key_query = SQL_STRINGS["read_pkey_table"][self.flavor]
        #  get conflicting pkey values before insert
        duplicate_val_before = self._get_exec().execute(
            duplicate_key_query, duplicate_keys
        )
        data_from_db_before = sorted(val[0] for val in duplicate_val_before)
```